packages/pyright-internal/src/analyzer/program.ts

```diff
@@ -240,6 +245,37 @@ export class Program {
         this._cacheManager.unregisterCacheOwner(this);
     }
 
+    enterEditMode() {
+        this._editMode = true;
+    }
+
+    leaveEditMode() {
+        this._editMode = false;
+        const edits: FileEditAction[] = [];
+        this._copiedOnWriteFiles.forEach((fileInfo, path) => {
+            const changes = this._copiedOnWriteChanges.get(path);
```

Comment:
Hmm, this is way more complicated than what I had in mind. I proposed that each SourceFile have an "isEditMode" flag and maintain its own copy-on-write file contents. That way, the program doesn't need to maintain any additional state.

Reply:
The SourceFileInfo needs to be recreated too I believe because of its list of imports. Since the SourceFileInfo doesn't have any code associated with it, I put it at this level.

Reply:
I don't see why we'd need to create a new SourceFileInfo. The list of imports is dictated by the contents of the file. If you change the contents of the file, it will be marked dirty, reparsed, and its imports will be recomputed as needed. It leverages all of the existing machinery in the "program" and "sourceFile" modules.

Reply:
Ah okay. Sounds good. Didn't know what updated the import list. I'll try moving it to the SourceFile then.

Reply:
Wait, I think this has to be at the program level because the source file may not exist. I could push the change tracking down to the SourceFile, but the list of changed source files has to be maintained in the program because it has to remove newly created files?

Reply:
Yes, you should be able to handle this in the SourceFile itself and leverage all of the other machinery in the program for tracking SourceFileInfos and their lifetimes.

Reply:
I moved the logic to the SourceFile. Not sure if it's simpler or not. Copying the state out of the SourceFile was more selective so it changed a lot more. Handling it is certainly less complicated though.